Rewrite UpdateStatus as an ES Module

kr8n3r opened the pull request:
We want to migrate all of our legacy Javascript files to ESM, so this is the latest in the series of conversions.

Rewrite follows the same conventions we previously used for FocusBanner, Collapsible checkboxes, ColourPreview, FileUpload, Autofocus, PreviewPane, CopyToClipboard, EnhancedTextbox and others.

Added bonus: removes the use of jQuery

**Fix of the tests authored by @tombye** 

In this PR we:

- rewrite the module as ES Module
- update the tests to be ES Module
- remove old module JS file
- remove old module from the legacy bundle build

## How to test

Edit SMS template page is the easiest as the "Will be charged as 1 text message" is updated using this script

kr8n3r committed:
Rewrite UpdateStatus as an ES Module

We want to migrate all of our legacy Javascript
files to ESM, so this is the latest in the series
of conversions.

Rewrite follows the same conventions we previously
used for FocusBanner, Collapsible checkboxes,
ColourPreview, FileUpload, Autofocus, PreviewPane,
CopyToClipboard and others.

We Fetch API to do teh posting instead of jQuery's
AJAX.

In this PR we:

- rewrite the module as ES Module
- update the tests to be ES Module
- remove old module JS file
- remove old module from the legacy bundle build

tombye committed:
Fix problems with testing async code

The tests run, and pass, but produce errors in the
console. After a bit of digging, this is mainly
due to the async bits of UpdateStatus code not
being handled properly. This aims to fix those
problems by:
- spying on updateStatus.update, to get
  access to the Promise it returns
- providing the chain of 2 Promises returned by
  window.fetch in all tests, to prevent errors in
  the code run when they eventually resolve
- converting all test functions to async, to
  enable use of the await operator inside them
- awaiting updateStatus.update at the end of the
  tests that don't do that already, to ensure all
  Promises returned by calls to it run their
  'resolve' code within the scope of the test

The last one might need some explanation. By
default, Promises always run their resolve or
reject callback code after the code that creates
them. Even if they're resolved immediately after
instantiation (like we do in the tests).

That's because JS runs in tasks and microtasks.
There's a queue for each to control the sequence
in which they're run.

The main program (the testsuite, when you're
running the tests) runs as a task. When it creates
new Promises and they resolve or reject, the
resulting callback is added to the microtasks
queue. By default, microtasks get dealt with after
the task completes. So when updateStatus.update is
called and its fetch creates a new Promise, the
code in updateStatus.update is affectively done
because the rest of it is in the resolve or reject
callbacks and they won't be run until it fulfills
and queues the winning callback as a microtask.

We use the await keyword as a nice way to mark the
points the updateStatus.update code waits for
Promises to fulfill. But it still returns control
to the code that called it when this happens and
that code will execute until it reaches the end of
the task.

In a browser it matters less when the callbacks
the updateStatus.update create run because it all
happens in milliseconds so users just see the end
result and the wait for all the tasks and
microtasks to complete is usually too small to
notice.

Our tests need to check the impact of running
updateStatus.update immediately after running it,
so we need a way to tell them to wait for all the
microtasks it creates to have been run. Awaiting
the Promise updateStatus.update returns means the
main task (aka the testsuite) stops, waits for it
to resolve and when it does it picks the microtask
that's been created for its resolve callback off
the queue. Because control is then moved back to
the main task (the testsuite), our assertions can
check the results of the callback.

But there's more: Control isn't passed back to the
main task (the testsuite) until all the microtasks
on the queue have been run. Because of this, all
Promises created by running updateStatus.update
will complete by the time control passes back to
the main task. So in tests that run
updateStatus.update multiple times, that extra
await at the end actually results in all the
Promises made completing.

More context on tasks and microtasks, and how
await moves controls between them, is in these
articles on MDN:
- https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide ('Tasks' and 'Microtasks' sections)
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#control_flow_effects_of_await

Extra: This commit also fixes a problem in the
UpdateStatus code, adding a .bind() to ensure the
throttle method is bound to the UpdateStatus
instance when used as an event handler.

tombye commented on tests/javascripts/update-status.test.mjs:
```diff
@@ -98,32 +88,39 @@ describe('Update content', () => {
-        "method": "post"
-      }
-    ]);
+    await updateStatus.init();
```
Resolved by my latest commit.

kr8n3r committed:
Replace custom getParent fucntion with .closest

kr8n3r committed:
Show parent node if hidden

Change made in https://github.com/alphagov/notifications-admin/commit/04aa3de4b6c554673665777ba3d4c8bb75a17ebc

kr8n3r reviewed (commented):
thorough write-up of what happens and why the change in the tests.

tombye reviewed (approved):
Thanks for making the changes, and catching the code to show the parent. Now looks good to me 👍🏻
